### PR TITLE
 [jit][edge] Migrate to TypeFactory for jit types on mobile.

### PR DESCRIPTION
**Pull request by zhxchen17**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71516

Function schema should be able to contruct dynamic types when on mobile.

Differential Revision: [D33673958](https://our.internmc.facebook.com/intern/diff/D33673958/)

### CODE REVIEW

File: torch/csrc/jit/serialization/export_module.cpp [hunk 1]
```diff
@@ -108,7 +108,11 @@ std::pair<IValue, IValue> getFunctionTuple(
   static const std::string torch_prefix("__torch__");
   static const std::string class_prefix("__torch__.torch.classes");
 
-  for (const TypePtr& t : mobile_code.types_) {
+  for (const TypePtr& ty : mobile_code.types_) {
+    auto t = ty;
+    if (auto dyn = t->castRaw<c10::DynamicType>()) {
+      t = dyn->fallback();
```
Comment (cccclai) [inline torch/csrc/jit/serialization/export_module.cpp:114]: Is it always the case? Or does it only apply to some types?

Comment (zhxchen17) [reply torch/csrc/jit/serialization/export_module.cpp:114]: @cccclai this branch is only taken when we call save_for_mobile, and load it back, and re-save it, which happens in flatbuffer and lite interpreter unit tests.
For the one major flow we support (server jit module - save -> mobile module), this branch is never taken.

Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp:114]: In the future, if we migrate to flatbuffer completely, the if statement will always be true?

**Pull request closed without merging**